docs/options_introduction.md
## Introduction to options
If you need more effects,  use '--option your-parameters' to enter what you need.

### Base

|    Option    |        Description         |                 Default                 |
| :----------: | :------------------------: | :-------------------------------------: |
|  --gpu_id   |   if -1, do not use gpu    |                    0                    |
| --media_path | your videos or images path |            ./imgs/ruoruo.jpg            |
| --start_time | start position of video, default is the beginning of video | '00:00:00' |
| --last_time | limit the duration of the video, default is the entire video | '00:00:00' |
|    --mode    |    program running mode(auto/clean/add/style)    |                 'auto'                  |
| --model_path |   pretrained model path    | ./pretrained_models/mosaic/add_face.pth |
| --result_dir |  output media will be saved here|                 ./result          |
| --temp_dir | Temporary files will go here | ./tmp |
|    --fps    |    read and output fps, if 0-> origin    |                 0                  |
| --no_preview | if specified,do not preview images when processing video. eg.(when run it on server) | Flase |

### AddMosaic

|    Option    |        Description         |                 Default                 |
| :----------: | :------------------------: | :-------------------------------------: |
| --mosaic_mod | type of mosaic -> squa_avg/ squa_random/ squa_avg_circle_edge/ rect_avg/random |                    squa_avg                    |
| --mosaic_size | mosaic size,if 0 -> auto size |            0            |
|    --mask_extend    |    extend mosaic area    |         10  |
| --mask_threshold | threshold of recognize mosaic position 0~255 | 64 |

### CleanMosaic

|    Option    |        Description         |                 Default                 |
| :----------: | :------------------------: | :-------------------------------------: |
| --traditional | if specified, use traditional image processing methods to clean mosaic |                                        |
| --tr_blur | ksize of blur when using traditional method, it will affect final quality |            10            |
|    --tr_down    |    downsample when using traditional method,it will affect final quality    |         10  |
| --medfilt_num | medfilt window of mosaic movement in the video | 11 |

### Style Transfer

|    Option    |        Description         |                 Default                 |
| :----------: | :------------------------: | :-------------------------------------: |
| --output_size | size of output media, if 0 -> origin |512|

docs/options_introduction_CN.md
## 参数说明
如果需要更多的效果,  请按照 '--option your-parameters' 输入所需要的参数

### 基本

|    选项    |        描述         |                 默认                 |
| :----------: | :------------------------: | :-------------------------------------: |
|  --gpu_id   |   if -1, do not use gpu    |                    0                    |
| --media_path | 需要处理的视频或者照片的路径 |            ./imgs/ruoruo.jpg            |
| --start_time | 视频开始处理的位置，默认从头开始 | '00:00:00' |
| --last_time | 处理的视频时长，默认是整个视频 | '00:00:00' |
|    --mode    |    运行模式(auto/clean/add/style)    |                 'auto'                  |
| --model_path |   预训练模型的路径    | ./pretrained_models/mosaic/add_face.pth |
| --result_dir | 保存路径 |                 ./result          |
| --temp_dir | 临时文件存储目录 | ./tmp |
|    --fps    |    限制视频输出的fps，0则为默认    |                 0                  |
| --no_preview | 如果输入，将不会在处理视频时播放实时预览．比如当你在服务器运行的时候 | Flase |

### 添加马赛克

|    选项    |        描述       |                 默认                 |
| :----------: | :------------------------: | :-------------------------------------: |
| --mosaic_mod | 马赛克类型 -> squa_avg/ squa_random/ squa_avg_circle_edge/ rect_avg/random |                    squa_avg                    |
| --mosaic_size | 马赛克大小，０则为自动 |            0            |
|    --mask_extend    |    拓展马赛克区域    |         10  |
| --mask_threshold | 马赛克区域识别阈值 0~255 | 64 |

### 去除马赛克

|    选项    |        描述       |                 默认                 |
| :----------: | :------------------------: | :-------------------------------------: |
| --traditional | 如果输入这个参数则使用传统方法清除马赛克 |                                        |
| --tr_blur | 传统方法模糊尺寸 |            10            |
|    --tr_down    |    传统方法下采样尺寸    |         10  |
| --medfilt_num | medfilt window of mosaic movement in the video | 11 |

### 风格转换

|    选项    |        描述       |                 默认                 |
| :----------: | :------------------------: | :-------------------------------------: |
| --output_size | 输出媒体的尺寸，如果是０则为原始尺寸 |512|

docs/Release_notes.txt
DeepMosaics V0.3.0
Core program building with windows10_1703_x86_64 
 + python 3.68  
 + pyinstaller 3.5
GUI building with C#
For more detail, please view on github: https://github.com/HypoX64/DeepMosaics

Releases History
 V0.3.0
   1. Support BiSeNet(Better recognition of mosaics).
   2. New videoHD model.
   3. Better feathering method.
 V0.2.0
   1. Add video model.
   2. Now you can input chinese path
   3. Support style transfer
   4. Support fps limit
 V0.1.2
   1. Support pix2pixHD model
 V0.1.1
   1. Check path, can't input illegal path
 V0.1.0
   1. Initial release.

cores/options.py
import argparse
import os
import sys


class Options():
    def __init__(self):
        self.parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
        self.initialized = False

    def initialize(self):

        #base
        self.parser.add_argument('--debug', action='store_true', help='if specified, start debug mode')
        self.parser.add_argument('--gpu_id', type=str,default='0', help='if -1, use cpu')
        self.parser.add_argument('--media_path', type=str, default='./imgs/ruoruo.jpg',help='your videos or images path')
        self.parser.add_argument('-ss', '--start_time', type=str, default='00:00:00',help='start position of video, default is the beginning of video')
        self.parser.add_argument('-t', '--last_time', type=str, default='00:00:00',help='duration of the video, default is the entire video')
        self.parser.add_argument('--mode', type=str, default='auto',help='Program running mode. auto | add | clean | style')
        self.parser.add_argument('--model_path', type=str, default='./pretrained_models/mosaic/add_face.pth',help='pretrained model path')
        self.parser.add_argument('--result_dir', type=str, default='./result',help='output media will be saved here')
        self.parser.add_argument('--temp_dir', type=str, default='./tmp', help='Temporary files will go here')
        self.parser.add_argument('--tempimage_type', type=str, default='jpg',help='type of temp image, png | jpg, png is better but occupy more storage space')
        self.parser.add_argument('--netG', type=str, default='auto',
            help='select model to use for netG(Clean mosaic and Transfer style) -> auto | unet_128 | unet_256 | resnet_9blocks | HD | video')
        self.parser.add_argument('--fps', type=int, default=0,help='read and output fps, if 0-> origin')
        self.parser.add_argument('--no_preview', action='store_true', help='if specified,do not preview images when processing video. eg.(when run it on server)')
        self.parser.add_argument('--output_size', type=int, default=0,help='size of output media, if 0 -> origin')
        self.parser.add_argument('--mask_threshold', type=int, default=64,help='threshold of recognize clean or add mosaic position 0~255')

        #AddMosaic
        self.parser.add_argument('--mosaic_mod', type=str, default='squa_avg',help='type of mosaic -> squa_avg | squa_random | squa_avg_circle_edge | rect_avg | random')
        self.parser.add_argument('--mosaic_size', type=int, default=0,help='mosaic size,if 0 auto size')
        self.parser.add_argument('--mask_extend', type=int, default=10,help='extend mosaic area')
        
        #CleanMosaic     
        self.parser.add_argument('--mosaic_position_model_path', type=str, default='auto',help='name of model use to find mosaic position')
        self.parser.add_argument('--traditional', action='store_true', help='if specified, use traditional image processing methods to clean mosaic')
        self.parser.add_argument('--tr_blur', type=int, default=10, help='ksize of blur when using traditional method, it will affect final quality')
        self.parser.add_argument('--tr_down', type=int, default=10, help='downsample when using traditional method,it will affect final quality')
        self.parser.add_argument('--no_feather', action='store_true', help='if specified, no edge feather and color correction, but run faster')
        self.parser.add_argument('--all_mosaic_area', action='store_true', help='if specified, find all mosaic area, else only find the largest area')
        self.parser.add_argument('--medfilt_num', type=int, default=11,help='medfilt window of mosaic movement in the video')
        self.parser.add_argument('--ex_mult', type=str, default='auto',help='mosaic area expansion')
        
        #StyleTransfer
        self.parser.add_argument('--preprocess', type=str, default='resize', help='resize and cropping of images at load time [ resize | resize_scale_width | edges | gray] or resize,edges(use comma to split)')
        self.parser.add_argument('--edges', action='store_true', help='if specified, use edges to generate pictures,(input_nc = 1)')  
        self.parser.add_argument('--canny', type=int, default=150,help='threshold of canny')
        self.parser.add_argument('--only_edges', action='store_true', help='if specified, output media will be edges')

        self.initialized = True


    def getparse(self, test_flag = False):
        if not self.initialized:
            self.initialize()
        self.opt = self.parser.parse_args()
        
        model_name = os.path.basename(self.opt.model_path)
        self.opt.temp_dir = os.path.join(self.opt.temp_dir, 'DeepMosaics_temp')

        if self.opt.gpu_id != '-1':
            os.environ["CUDA_VISIBLE_DEVICES"] = str(self.opt.gpu_id)
            import torch
            if not torch.cuda.is_available():
                self.opt.gpu_id = '-1'
        # else:
        #     self.opt.gpu_id = '-1'

        if test_flag:
            if not os.path.exists(self.opt.media_path):
                print('Error: Media does not exist!')
                input('Please press any key to exit.\n')
                sys.exit(0)
            if not os.path.exists(self.opt.model_path):
                print('Error: Model does not exist!')
                input('Please press any key to exit.\n')
                sys.exit(0)

            if self.opt.mode == 'auto':
                if 'clean' in model_name or self.opt.traditional:
                    self.opt.mode = 'clean'
                elif 'add' in model_name:
                    self.opt.mode = 'add'
                elif 'style' in model_name or 'edges' in model_name:
                    self.opt.mode = 'style'
                else:
                    print('Please check model_path!')
                    input('Please press any key to exit.\n')
                    sys.exit(0)

            if self.opt.output_size == 0 and self.opt.mode == 'style':
                self.opt.output_size = 512

            if 'edges' in model_name or 'edges' in self.opt.preprocess:
                self.opt.edges = True

            if self.opt.netG == 'auto' and self.opt.mode =='clean':
                if 'unet_128' in model_name:
                    self.opt.netG = 'unet_128'
                elif 'resnet_9blocks' in model_name:
                    self.opt.netG = 'resnet_9blocks'
                elif 'HD' in model_name and 'video' not in model_name:
                    self.opt.netG = 'HD'
                elif 'video' in model_name:
                    self.opt.netG = 'video'
                else:
                    print('Type of Generator error!')
                    input('Please press any key to exit.\n')
                    sys.exit(0)

            if self.opt.ex_mult == 'auto':
                if 'face' in model_name:
                    self.opt.ex_mult = 1.1
                else:
                    self.opt.ex_mult = 1.5
            else:
                self.opt.ex_mult = float(self.opt.ex_mult)

            if self.opt.mosaic_position_model_path == 'auto' and self.opt.mode == 'clean':
                _path = os.path.join(os.path.split(self.opt.model_path)[0],'mosaic_position.pth')
                if os.path.isfile(_path):
                    self.opt.mosaic_position_model_path = _path
                else:
                    input('Please check mosaic_position_model_path!')
                    input('Please press any key to exit.\n')
                    sys.exit(0)

        return self.opt